core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java

```diff
@@ -376,5 +409,105 @@ public Pair<Integer, TimeUnit> refresh(RESTClient client) {
 
       return null;
     }
+
+    @SuppressWarnings("FutureReturnValueIgnored")
+    public static void scheduleTokenRefresh(
+        RESTClient client,
+        ScheduledExecutorService tokenRefreshExecutor,
+        AuthSession session,
+        long startTimeMillis,
+        long expiresIn,
+        TimeUnit unit) {
+      // convert expiration interval to milliseconds
+      long expiresInMillis = unit.toMillis(expiresIn);
+      // how much ahead of time to start the request to allow it to complete
+      long refreshWindowMillis = Math.min(expiresInMillis / 10, MAX_REFRESH_WINDOW_MILLIS);
+      // how much time to wait before expiration
+      long waitIntervalMillis = expiresInMillis - refreshWindowMillis;
+      // how much time has already elapsed since the new token was issued
+      long elapsedMillis = System.currentTimeMillis() - startTimeMillis;
+      // how much time to actually wait
+      long timeToWait = Math.max(waitIntervalMillis - elapsedMillis, MIN_REFRESH_WAIT_MILLIS);
+
+      tokenRefreshExecutor.schedule(
+          () -> {
+            long refreshStartTime = System.currentTimeMillis();
+            Pair<Integer, TimeUnit> expiration = session.refresh(client);
```

Comment:
The refresh done here also needs to handle cases where the token has expired.

Reply:
what specifically did you have in mind here? I looked at the logic in https://github.com/apache/iceberg/blob/3f01ed874bea66a05195e491c9b7045a4337d902/core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java#L368 and it seems to do the right thing (as it relies on the expiration time that is being part of the response)

Reply:
I'm referring to retries. If the refresh fails because the token has expired, then this should retry using basic auth, if possible.

Reply:
I've pushed something where we try and refresh the token using the credential when the normal retries fail. Let me know if this is what you had in mind